Plutarch/BitString.hs

```diff
@@ -0,0 +1,147 @@
+module Plutarch.BitString (
+  -- * Type
+  PBitString (..),
+
+  -- * Functions
+  preadBit,
+  psetBits,
+  pclearBits,
+  pshift,
+  protate,
+  pcountSetBits,
+  pfindFirstSetBit,
+  pfindFirstSetBit',
+) where
+
+import GHC.Generics (Generic)
+import Plutarch.Bool (PBool, PEq, POrd, PPartialOrd, pif, (#<))
+import Plutarch.Builtin (PBuiltinList)
+import Plutarch.ByteString (PByteString)
+import Plutarch.Integer (PInteger)
+import Plutarch.Internal (
+  S,
+  Term,
+  phoistAcyclic,
+  plet,
+  punsafeBuiltin,
+  (#),
+  (:-->),
+ )
+import Plutarch.Internal.Newtype (PlutusTypeNewtype)
+import Plutarch.Internal.Other (pto)
+import Plutarch.Internal.PLam (plam)
+import Plutarch.Internal.PlutusType (
+  DerivePlutusType (DPTStrat),
+  PlutusType,
+  pcon,
+ )
+import Plutarch.Lift (pconstant)
+import Plutarch.Maybe (PMaybe (PJust, PNothing))
+import PlutusCore qualified as PLC
+
+{- | A wrapper around 'PByteString' for CIP-122 and CIP-123 bitwise operations.
+
+= Note
+
+This type exists because /bit/ and /byte/ indexes work in different
+directions. To avoid confusing behaviour, we require an explicit wrapping of
+'PByteString's to use bitwise functionality: this way, it's clear where which
+scheme applies.
+
+@since WIP
+-}
+newtype PBitString (s :: S) = PBitString (Term s PByteString)
+  deriving stock
+    ( -- | @since WIP
+      Generic
+    )
+  deriving anyclass
+    ( -- | @since WIP
+      PlutusType
+    , -- | @since WIP
+      PEq
+    , -- | @since WIP
+      PPartialOrd
+    , -- | @since WIP
+      POrd
+    )
+
+-- | @since WIP
+instance DerivePlutusType PBitString where
+  type DPTStrat _ = PlutusTypeNewtype
+
+{- | Bit access operation, as defined in [CIP-122](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0122#readbit).
+
+@since WIP
+-}
+preadBit :: forall (s :: S). Term s (PBitString :--> PInteger :--> PBool)
+preadBit = plam $ \bs i -> punsafeBuiltin PLC.ReadBit # pto bs # i
+
+{- | Given a list of positions, set the bits at those positions.
+
+This works similarly to the @writeBits@ operation in
+[CIP-122](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0122#writebits)
+with regard to the list of indexes. Effectively, @psetBits b ixes@ is
+equivalent to @writeBits b . map (, True) $ ixes@. All caveats that this
+entails from the CIP-122 description apply.
+
+@since WIP
+-}
+psetBits :: forall (s :: S). Term s (PBitString :--> PBuiltinList PInteger :--> PBitString)
+psetBits = plam $ \bs ixes -> punsafeBuiltin PLC.WriteBits # pto bs # ixes # pconstant True
+
+{- | As 'psetBits', but clears the bits instead.
+
+@since WIP
+-}
+pclearBits :: forall (s :: S). Term s (PBitString :--> PBuiltinList PInteger :--> PBitString)
+pclearBits = plam $ \bs ixes -> punsafeBuiltin PLC.WriteBits # pto bs # ixes # pconstant False
```

Comment:
What would be rational to have this separately as functions? As above, just `punsafeBuiltin`ing is more efficient. We still get more typelevel assurance as it can still have `PBitString` type.